File: chess-app/src/engine/Board.js
import Pawn from './pieces/Pawn.js';
import Rook from './pieces/Rook.js';
import Knight from './pieces/Knight.js';
import Bishop from './pieces/Bishop.js';
import Queen from './pieces/Queen.js';
import King from './pieces/King.js'


export default class ChessBoard {
    constructor() {
        this.board = this._initializeEmptyBoard();
        this.whiteKingPos = [];
        this.blackKingPos = [];
        this.whitePieces = [];
        this.blackPieces = [];
        this.castlingRights = {
            white: {kingside: true, queenside: true},
            black: {kingside: true, queenside: true}
        }
        this.enPassantTarget = null;
    }

    deepCopy() {
        const copy = new ChessBoard();
        for (let i = 0; i < 8; i++){
            for (let j = 0; j < 8; j++){
                if (this.board[i][j] != null) {
                    copy.board[i][j] = this.board[i][j].deepCopy();
                }
            }
        }
        copy.whiteKingPos = this.whiteKingPos.slice();
        copy.blackKingPos = this.blackKingPos.slice();
        if (this.enPassantTarget != null) {
            copy.enPassantTarget = this.enPassantTarget.slice();
        } else {
            copy.enPassantTarget = this.enPassantTarget
        }
    
        copy.whitePieces = this.whitePieces.map(coord => coord.slice());
        copy.blackPieces = this.blackPieces.map(coord => coord.slice());

        copy.castlingRights = JSON.parse(JSON.stringify(this.castlingRights));
        
        return copy;
    }

    _initializeEmptyBoard(){
        let board = [];
        for (let i = 0; i < 8; i++) {
            let row = [];
            for (let j = 0; j < 8; j++) {
                row.push(null);
            }
            board.push(row);
        }
        return board;
    }

    setNewBoard(){
        for (let i = 0; i < 8; i++) {
            this.board[6][i] = new Pawn('white');
            this.board[1][i] = new Pawn('black');
            this.whitePieces.push([6,i]);
            this.blackPieces.push([1,i]);
        }
        let i = 0;
        let j = 7;
        while (i < j) {
            if (i === j-1) {
                this.board[7][i] = new Queen('white');
                this.board[7][j] = new King('white');
                this.board[0][i] = new Queen('black');
                this.board[0][j] = new King('black');
                this.whiteKingPos = [7,j];
                this.blackKingPos = [0,j];
            }
            else {
                switch (i) {
                    case 0:
                        this.board[7][i] = new Rook('white');
                        this.board[7][j] = new Rook('white');
                        this.board[0][i] = new Rook('black');
                        this.board[0][j] = new Rook('black');
                        break;
                    case 1:
                        this.board[7][i] = new Knight('white');
                        this.board[7][j] = new Knight('white');
                        this.board[0][i] = new Knight('black');
                        this.board[0][j] = new Knight('black');
                        break;
                    case 2:
                        this.board[7][i] = new Bishop('white');
                        this.board[7][j] = new Bishop('white');
                        this.board[0][i] = new Bishop('black');
                        this.board[0][j] = new Bishop('black');
                        break;
                }
            }
            this.whitePieces.push([7,i]);
            this.whitePieces.push([7,j]);
            this.blackPieces.push([0,i]);
            this.blackPieces.push([0,j]);
            i++;
            j--;
        }
        return this.board;
    }

    _getPieceType(piece) {
        if (piece instanceof Pawn && piece.color === 'white') return 'wP';
        if (piece instanceof Pawn && piece.color === 'black') return 'bP';
        if (piece instanceof Rook && piece.color === 'white') return 'wR';
        if (piece instanceof Rook && piece.color === 'black') return 'bR';
        if (piece instanceof Knight && piece.color === 'white') return 'wN';
        if (piece instanceof Knight && piece.color === 'black') return 'bN';
        if (piece instanceof Bishop && piece.color === 'white') return 'wB';
        if (piece instanceof Bishop && piece.color === 'black') return 'bB';
        if (piece instanceof Queen && piece.color === 'white') return 'wQ';
        if (piece instanceof Queen && piece.color === 'black') return 'bQ';
        if (piece instanceof King && piece.color === 'white') return 'wK';
        if (piece instanceof King && piece.color === 'black') return 'bK';
        return null;
    }

    getBoard() {
        let board = [];
        for (let i = 0; i < 8; i++) {
            let row = [];
            for (let j = 0; j < 8; j++) {
                if (this.board[i][j] === null) {
                    row.push('');
                } else {
                    row.push(this._getPieceType(this.board[i][j]));
                }
            }
            board.push(row);
        }
        return board;
    }

    getPieceAt(position) {
        const [row, col] = position;
        if (this.isInsideBoard(position) === false) {
            throw new Error('Position out of bounds');
        }
        return this.board[row][col];
    }

    setPieceAt(position, piece) {
        const [row, col] = position;
        if (this.isInsideBoard(position) === false) {
            throw new Error('Position out of bounds');
        }
        this.board[row][col] = piece;
    }

    isInsideBoard(position) {
        const [row, col] = position;
        return row >= 0 && row < 8 && col >= 0 && col < 8;
    }

    movePiece(fromPos, toPos) {
        if (!this.isInsideBoard(fromPos) || !this.isInsideBoard(toPos)) {
            throw new Error('Position out of bounds')
        }
        let [fR, fC] = fromPos
        let [tR, tC] = toPos
        let piece = this.board[fR][fC]
        piece.hasMoved = true
        this.board[tR][tC] = piece
    }

    getAllPiecesColor(color){
        switch(color){
            case 'white':
                return this.whitePieces;
            case 'black':
                return this.blackPieces;
        }
    }

    getAllMovesColor(color){
        let moves = {};
        let pieceList = [];
        if(color === 'white'){
            pieceList = this.whitePieces;
        } else {
            pieceList = this.blackPieces;
        }
        pieceList.forEach((piece)=>{
            const currPiece = this.getPieceAt(piece);
            moves[piece] = currPiece.getPseudoLegalMoves(this, piece)
            // moves.push(currPiece.getPseudoLegalMoves())
        })
        return moves
    }

    isCheck(enemyPieceList, kingPos) {

        for(let i=0; i<enemyPieceList.length; i++) {
            const pos = enemyPieceList[i];
            const currPiece = this.getPieceAt(pos);
            const pseudoLegalMoves = currPiece.getPseudoLegalMoves(this, pos, true)
            const isKingIncluded = pseudoLegalMoves.some(arr => JSON.stringify(arr) === JSON.stringify(kingPos));

            if (isKingIncluded) {
                return true;
            }
        }
        return false;
    }

    canCastle(color, side) {
        const betweenPositions = {
            white: {kingside: [[7,5], [7,6]], queenside: [[7,1], [7,2], [7, 3]]},
            black: {kingside: [[0,5], [0,6]], queenside: [[0,1], [0,2], [0, 3]]},
        }
        const castlingRight = this.castlingRights[color][side];
        if (!castlingRight) {
            return false;
        }

        const enemyPieceList = color === 'white' ? this.blackPieces : this.whitePieces;
        let kingPos = color === 'white' ? this.whiteKingPos : this.blackKingPos;
        if (this.isAttacked(enemyPieceList, kingPos)) {
            return false;
        }

        const bwPos = betweenPositions[color][side]
        for (let i=0; i < bwPos.length; i++) {
            if (this.getPieceAt(bwPos[i]) !== null) {
                return false;
            }
        }

        for (let i=0; i<2; i++) {
            if (this.isAttacked(enemyPieceList, bwPos[i])) {
                return false;
            }
        }

        return true;
    }

    isAttacked(enemyPieceList, kingPos) {

        for(let i=0; i<enemyPieceList.length; i++) {
            const pos = enemyPieceList[i];
            const currPiece = this.getPieceAt(pos);
            const pseudoLegalMoves = currPiece.getPseudoLegalMoves(this, pos, true)
            const isKingIncluded = pseudoLegalMoves.some(arr => JSON.stringify(arr) === JSON.stringify(kingPos));

            if (isKingIncluded) {
                return true;
            }
        }
        return false;
    }

    // ----------------------------------ADDITIONAL FNS------------------------------------//
    setNewPiece(row, col, type, color) {
        console.log(`Setting new piece: ${type} at (${row}, ${col}) with color ${color}`);
        switch(type){
            case 'Rook':
                this.board[row][col] = new Rook(color)
                break;
            case 'King':
                this.board[row][col] = new King(color)
                break;
            case 'Queen':
                this.board[row][col] = new Queen(color)
                break;
            case 'Bishop':
                this.board[row][col] = new Bishop(color)
                break;
            case 'Knight':
                this.board[row][col] = new Knight(color)
                break;
            case 'Pawn':
                this.board[row][col] = new Pawn(color)
                break;
            default:
                throw new Error('Invalid piece type');
        }
    }
    
    resetBoard() {
        this.board = this.createBoard();
    }

    getMoves(r,c){
        return this.board[r][c].getPseudoLegalMoves(this, [r,c])
    }

    printBoard() {
        for (let i = 0; i < 8; i++) {
            let row = '';
            for (let j = 0; j < this.board[i].length; j++) {
                if (this.board[i][j] === null) {
                    row += ' . ';
                } else {
                    if (this.board[i][j].color === 'white'){
                        row += ` ${this.board[i][j].type[0].toUpperCase()} `;
                    } else {
                        row += ` ${this.board[i][j].type[0]} `;
                    }
                }
            }
            console.log(row);
        }
    }
}

File: chess/src/main/java/com/app/chess/Board.java
package com.app.chess;

import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashMap;
import com.app.chess.Pieces.*;

public class Board {
    Piece[][] board;
    int[] whiteKingPos;
    int[] blackKingPos;
    List<int[]> whitePieces;
    List<int[]> blackPieces;
    HashMap<Player, HashMap<String, Boolean>> castlingRights;
    int[] enPassantTarget;

    public Board() {
        board = initializeEmptyBoard();
        castlingRights.keySet().addAll(Arrays.asList(Player.WHITE, Player.BLACK));
        castlingRights.get(Player.WHITE).put("kingside", true);
        castlingRights.get(Player.WHITE).put("queenside", true);
        castlingRights.get(Player.BLACK).put("kingside", true);
        castlingRights.get(Player.BLACK).put("queenside", true);
        enPassantTarget = null;
    }

    public Board deepCopy() {
        Board newBoard = new Board();
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                if (this.board[i][j] != null) {
                    newBoard.board[i][j] = this.board[i][j].deepCopy();
                } else {
                    newBoard.board[i][j] = null;
                }
            }
        }
        newBoard.whiteKingPos = Arrays.copyOf(this.whiteKingPos, this.whiteKingPos.length);
        newBoard.blackKingPos = Arrays.copyOf(this.blackKingPos, this.blackKingPos.length);
        newBoard.whitePieces = new ArrayList<>();
        for (int[] pos : this.whitePieces) {
            newBoard.whitePieces.add(Arrays.copyOf(pos, pos.length));
        }
        newBoard.blackPieces = new ArrayList<>();
        for (int[] pos : this.blackPieces) {
            newBoard.blackPieces.add(Arrays.copyOf(pos, pos.length));
        }
        newBoard.castlingRights = new HashMap<>();
        for (Player player : this.castlingRights.keySet()) {
            HashMap<String, Boolean> rights = new HashMap<>(this.castlingRights.get(player));
            newBoard.castlingRights.put(player, rights);
        }
        if (this.enPassantTarget != null) {
            newBoard.enPassantTarget = Arrays.copyOf(this.enPassantTarget, this.enPassantTarget.length);
        } else {
            newBoard.enPassantTarget = null;
        }
        return newBoard;
    }

    private Piece[][] initializeEmptyBoard() {
        Piece[][] newBoard = new Piece[8][8];
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                newBoard[i][j] = null;
            }
        }
        return newBoard;
    }

    public void setNewBoard() {
        // A simple array of pieces to place
        Piece[] backRow = new Piece[]{
            new Rook(), new Knight(), new Bishop(), new Queen(),
            new King(), new Bishop(), new Knight(), new Rook()
        };

        // Simplified and more readable setup
        for (int i = 0; i < 8; i++) {
            // Pawns
            board[6][i] = new Pawn(Player.WHITE);
            board[1][i] = new Pawn(Player.BLACK);
            whitePieces.add(new int[]{6, i});
            blackPieces.add(new int[]{1, i});

            // Back Row
            board[7][i] = backRow[i].setPlayer(Player.WHITE);
            board[0][i] = backRow[i].setPlayer(Player.BLACK);
            whitePieces.add(new int[]{7, i});
            blackPieces.add(new int[]{0, i});
        }

        // Set king positions separately for clarity
        whiteKingPos = new int[]{7, 4};
        blackKingPos = new int[]{0, 4};
    }

    private String getPieceType(Piece piece) {
        if (piece instanceof King) {
            return piece.color == Player.WHITE ? "wK" : "bK";
        } else if (piece instanceof Queen) {
            return piece.color == Player.WHITE ? "wQ" : "bQ";
        } else if (piece instanceof Rook) {
            return piece.color == Player.WHITE ? "wR" : "bR";
        } else if (piece instanceof Bishop) {
            return piece.color == Player.WHITE ? "wB" : "bB";
        } else if (piece instanceof Knight) {
            return piece.color == Player.WHITE ? "wN" : "bN";
        } else if (piece instanceof Pawn) {
            return piece.color == Player.WHITE ? "wP" : "bP";
        }
    }

    public String[][] getBoard() {
        String[][] displayBoard = new String[8][8];
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                if (board[i][j] == null) {
                    displayBoard[i][j] = "";
                } else {
                    displayBoard[i][j] = getPieceType(board[i][j]);
                }
            }
        }
        return displayBoard;
    }

    public Piece getPieceAt(int[] position) {
        if (isInsideBoard(position) == false) {
            System.out.println("Position out of board bounds");
            return null;
        }
        return board[position[0]][position[1]];
    }

    public void setPieceAt(int[] position, Piece piece) {
        if (isInsideBoard(position)) {
            board[position[0]][position[1]] = piece;
        }
    }

    public boolean isInsideBoard(int[] position) {
        return position[0] >= 0 && position[0] < 8 && position[1] >= 0 && position[1] < 8;
    }

    public void movePiece(int[] fromPos, int[] toPos) {
        if (isInsideBoard(fromPos) == false || isInsideBoard(toPos) == false) {
            System.out.println("Move out of board bounds");
            return;
        }
        Piece piece = board[fromPos[0]][fromPos[1]];
        piece.hasMoved = true;
        board[toPos[0]][toPos[1]] = piece;
    }

    public List<int[]> getAllPiecesPositions(Player color) {
        return color == Player.WHITE ? whitePieces : blackPieces;
    }

    public HashMap<Piece, List<int[]>> getAllMovesColor(Player color) {
        HashMap<Piece, List<int[]>> allMoves = new HashMap<>();
        List<int[]> piecesPositions = getAllPiecesPositions(color);
        for (int[] pos : piecesPositions) {
            Piece piece = getPieceAt(pos);
            if (piece != null) {
                List<int[]> pieceMoves = piece.getPsuedoLegalMoves(this, pos);
                allMoves.put(piece, pieceMoves);
            }
        }
        return allMoves;
    }

    public boolean isCheck(List<int []> enemyPieceList, int[] kingPos) {
        for (int[] pos : enemyPieceList) {
            Piece enemyPiece = getPieceAt(pos);
            if (enemyPiece != null) {
                List<int[]> enemyMoves = enemyPiece.getPsuedoLegalMoves(this, pos, true);
                for (int[] move : enemyMoves) {
                    if (Arrays.equals(move, kingPos)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    public boolean canCastle(Player color, String side) {
        HashMap<Player, HashMap<String, List<int[]>>> betweenPositions;
        betweenPositions = new HashMap<>() {{
            put(Player.WHITE, new HashMap<>() {{
                put("kingside", Arrays.asList(new int[]{7, 5}, new int[]{7, 6}));
                put("queenside", Arrays.asList(new int[]{7, 1}, new int[]{7, 2}, new int[]{7, 3}));
            }});
            put(Player.BLACK, new HashMap<>() {{
                put("kingside", Arrays.asList(new int[]{0, 5}, new int[]{0, 6}));
                put("queenside", Arrays.asList(new int[]{0, 1}, new int[]{0, 2}, new int[]{0, 3}));
            }});
        }};
        boolean rights = castlingRights.get(color).get(side);
        if (!rights) return false;

        List<int []> enemyPieceList = color == Player.WHITE ? blackPieces : whitePieces;
        int[] kingPos = color == Player.WHITE ? whiteKingPos : blackKingPos;
        if (isAttacked(enemyPieceList, kingPos)) return false;

        List<int[]> betweenPos = betweenPositions.get(color).get(side);
        for (int[] pos : betweenPos) {
            if (getPieceAt(pos) != null || isAttacked(enemyPieceList, pos)) {
                return false;
            }
        }
        return true;
    }

    public boolean isAttacked(List<int []> enemyPieceList, int[] position) {
        for (int[] pos : enemyPieceList) {
            Piece enemyPiece = getPieceAt(pos);
            if (enemyPiece != null) {
                List<int[]> enemyMoves = enemyPiece.getPsuedoLegalMoves(this, pos, true);
                for (int[] move : enemyMoves) {
                    if (Arrays.equals(move, position)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    public void setNewPiece(int r, int c, String pieceType, Player color) {
        Piece newPiece = null;
        switch (pieceType) {
            case "Queen":
                newPiece = new Queen(color);
                break;
            case "Rook":
                newPiece = new Rook(color);
                break;
            case "Bishop":
                newPiece = new Bishop(color);
                break;
            case "Knight":
                newPiece = new Knight(color);
                break;
            default:
                System.out.println("Invalid piece type for promotion");
                return;
        }
        board[r][c] = newPiece;
        List<int[]> pieceList = color == Player.WHITE ? whitePieces : blackPieces;
        pieceList.add(new int[]{r, c});
    }

    public void resetBoardState() {
        enPassantTarget = null;
        castlingRights.get(Player.WHITE).put("kingside", true);
        castlingRights.get(Player.WHITE).put("queenside", true);
        castlingRights.get(Player.BLACK).put("kingside", true);
        castlingRights.get(Player.BLACK).put("queenside", true);
        whitePieces.clear();
        blackPieces.clear();
        setNewBoard();
    }

    public List<int []> getMoves(int r, int c) {
        Piece piece = getPieceAt(new int[]{r, c});
        if (piece == null) {
            return null;
        }
        return piece.getPsuedoLegalMoves(this, new int[]{r, c});
    }

    public void printBoard() {
        String[][] displayBoard = getBoard();
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                System.out.print((displayBoard[i][j].isEmpty() ? "--" : displayBoard[i][j]) + " ");
            }
            System.out.println();
        }
        System.out.println();
    }
}
